File: android-lister/app/src/main/java/com/cloudstore/lister/domain/ListSearchLogic.kt
package com.cloudstore.lister.domain

import com.cloudstore.lister.data.InventoryListItem

object ListSearchLogic {
    /** Case-insensitive substring match across all searchable [InventoryListItem] fields. */
    fun filtered(items: List<InventoryListItem>, query: String): List<InventoryListItem> {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return items
        return items.filter { matches(it, trimmed) }
    }

    fun matches(item: InventoryListItem, query: String): Boolean {
        val needle = query.trim()
        if (needle.isEmpty()) return true
        return searchableValues(item).any { it.contains(needle, ignoreCase = true) }
    }

    private fun searchableValues(item: InventoryListItem): List<String> {
        return listOfNotNull(
            item.name.takeIf { it.isNotEmpty() },
            item.productType?.takeIf { it.isNotEmpty() },
            item.manufacturer?.takeIf { it.isNotEmpty() },
            item.barcode?.takeIf { it.isNotEmpty() },
            item.productId.toString(),
            item.priceLabel.takeIf { it.isNotEmpty() },
            item.stockLabel.takeIf { it.isNotEmpty() },
            item.pullCount.toString(),
        )
    }
}

File: ios-lister/CloudStoreLister/Logic/ListSearchLogic.swift
import Foundation

enum ListSearchLogic {
    /// Case-insensitive substring match across all searchable `InventoryListItem` fields.
    /// Empty / whitespace-only query returns `items` unchanged.
    static func filtered(_ items: [InventoryListItem], query: String) -> [InventoryListItem] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return items }
        return items.filter { matches($0, query: trimmed) }
    }

    static func matches(_ item: InventoryListItem, query: String) -> Bool {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !needle.isEmpty else { return true }
        return searchableValues(for: item).contains { $0.localizedCaseInsensitiveContains(needle) }
    }

    private static func searchableValues(for item: InventoryListItem) -> [String] {
        [
            item.name,
            item.productType,
            item.manufacturer,
            item.barcode,
            String(item.productId),
            item.priceLabel,
            item.stockLabel,
            String(item.pullCount),
        ]
        .compactMap { $0 }
        .filter { !$0.isEmpty }
    }
}
